docs/sql-migration-guide-upgrade.md

```diff
@@ -45,6 +45,8 @@ displayTitle: Spark SQL Upgrading Guide
 
   - In Spark version 2.4 and earlier, if `org.apache.spark.sql.functions.udf(Any, DataType)` gets a Scala closure with primitive-type argument, the returned UDF will return null if the input values is null. Since Spark 3.0, the UDF will return the default value of the Java type if the input value is null. For example, `val f = udf((x: Int) => x, IntegerType)`, `f($"x")` will return null in Spark 2.4 and earlier if column `x` is null, and return 0 in Spark 3.0. This behavior change is introduced because Spark 3.0 is built with Scala 2.12 by default.
 
+  - Since Spark 3.0, `DataFrame.toJSON()` in PySpark returns `DataFrame` of JSON string instead of `RDD`. The method in Scala/Java was changed to return `DataFrame` before, but the one in PySpark was not changed at that time. If you still want to return `RDD`, you can restore the previous behavior by setting `spark.sql.legacy.pyspark.toJsonShouldReturnDataFrame` to `false`.
```

Comment:
Backing up ... `Dataset` in Scala doesn't return a `Dataset[Row]` (i.e. `DataFrame`). It returns `Dataset[String]`. It already seems consistent. This isn't making it consistent then, but just an API change. That's possible of course, but is it really important? especially if it introduces a new flag, that toggles Pyspark behavior between what's consistent and inconsistent with Scala. (We can't have the flag change the return type in Scala dynamically)

Reply:
@ueshin what do you think of this argument, that it's already consistent?

Reply:
Actually I'm still feeling it's inconsistent because the abstraction layer is different between RDD and DataFrame/Dataset. I guess users expect it returns something of the same abstraction, i.e., RDD returns RDD, DataFrame returns DataFrame, so I'd rather handle Dataset as DataFrame in Python than RDD.We might still need to discuss the function `map` or the behavior of `DataFrameReader.csv`/`.json` as @HyukjinKwon mentioned.

Reply:
The current Scala method does _not_ return a DataFrame. That's where I'm confused about this argument. Yes, either that has to change too, or this change actually makes it less consistent.